smartsim/entity/ensemble.py

```diff
@@ -298,3 +307,165 @@ def _read_model_parameters(self):
+        orchestrator (converged or not) prior to the execution
+        of every entity belonging to this ensemble
+
+        Device selection is either "GPU" or "CPU". If many devices are
+        present, a number can be passed for specification e.g. "GPU:1".
+
+        Setting ``devices_per_node=N``, with N greater than one will result
+        in the model being stored in the first N devices of type ``device``.
+
+        One of either script (in memory string representation) or script_path (file)
+        must be provided
+
+        :param name: key to store script under
+        :type name: str
+        :param script: TorchScript code
+        :type script: str, optional
+        :param script_path: path to TorchScript code
+        :type script_path: str, optional
+        :param device: device for script execution, defaults to "CPU"
+        :type device: str, optional
+        :param devices_per_node: number of devices on each host
+        :type devices_per_node: int
+        """
+        db_script = DBScript(
+            name=name,
+            script=script,
+            script_path=script_path,
+            device=device,
+            devices_per_node=devices_per_node
+        )
+        self._db_scripts.append(db_script)
+        for entity in self:
+            self._extend_entity_db_scripts(entity, [db_script])
+
+    
+    def add_function(self, name, function=None, device="CPU", devices_per_node=1):
+        """TorchScript function to launch with every entity belonging to this ensemble
+
+        Each script function to the model will be loaded into a
+        non-converged orchestrator prior to the execution
+        of every entity belonging to this ensemble.
+        
+        For converged orchestrators, the :meth:`add_script` method should be used.
+
+        Device selection is either "GPU" or "CPU". If many devices are
+        present, a number can be passed for specification e.g. "GPU:1".
+
+        Setting ``devices_per_node=N``, with N greater than one will result
+        in the model being stored in the first N devices of type ``device``.
+
+        :param name: key to store function under
+        :type name: str
+        :param script: TorchScript code
+        :type script: str, optional
+        :param script_path: path to TorchScript code
+        :type script_path: str, optional
+        :param device: device for script execution, defaults to "CPU"
+        :type device: str, optional
+        :param devices_per_node: number of devices on each host
+        :type devices_per_node: int
+        """
+        db_script = DBScript(
+            name=name,
+            script=function,
+            device=device,
+            devices_per_node=devices_per_node
+        )
+        self._db_scripts.append(db_script)
+        for entity in self:
+            self._extend_entity_db_scripts(entity, [db_script])
+
+    def _extend_entity_db_models(self, model, db_models):
```

Comment:
why aren't we just calling ``add_db_model`` here so that we don't have to do this error handling twice?

Reply:
I wanted to avoid creating copies of the same `DBModel` or `DBScript`, and just add references to the lists (this makes it also easier to check for duplicates). Maybe I can add a `_append_db_model` function in `Model`, in which I catch the exception, and call it from `add_db_model` both for `Ensemble` and `Model`. Same for script. That way, the try-catch will only happen in one location.